lib/evmmax/evmmax.cpp

```diff
@@ -0,0 +1,114 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2023 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <evmmax/evmmax.hpp>
+
+using namespace intx;
+
+namespace evmmax
+{
+namespace
+{
+inline constexpr uint64_t mul_inv64(uint64_t base) noexcept
+{
+    uint64_t result = 1;
+    for (auto i = 0; i < 64; ++i)
+    {
+        result *= base;
+        base *= base;
+    }
+    return result;
+}
+
+template <typename UintT>
+inline UintT compute_r_squared(const UintT& mod) noexcept
+{
+    static constexpr auto r2 = intx::uint<UintT::num_bits * 2 + 64>{1} << (UintT::num_bits * 2);
+    return intx::udivrem(r2, mod).rem;
+}
+
+inline constexpr std::pair<uint64_t, uint64_t> addmul(
+    uint64_t t, uint64_t a, uint64_t b, uint64_t c) noexcept
+{
+    const auto p = umul(a, b) + t + c;
+    return {p[1], p[0]};
+}
+}  // namespace
+
+template <typename UintT>
+ModArith<UintT>::ModArith(const UintT& modulus) noexcept
+  : mod{modulus}, m_r_squared{compute_r_squared(modulus)}, m_mod_inv{mul_inv64(0 - modulus[0])}
+{}
+
+template <typename UintT>
+UintT ModArith<UintT>::mul(const UintT& x, const UintT& y) const noexcept
+{
+    // Coarsely Integrated Operand Scanning (CIOS) Method
+    // Based on 2.3.2 from
+    // High-Speed Algorithms & Architectures For Number-Theoretic Cryptosystems
+    // https://www.microsoft.com/en-us/research/wp-content/uploads/1998/06/97Acar.pdf
+
+    static constexpr auto S = UintT::num_words;
+
+    intx::uint<UintT::num_bits + 64> t;
+    for (size_t i = 0; i != S; ++i)
+    {
+        uint64_t c = 0;
+        for (size_t j = 0; j != S; ++j)
+        {
+            std::tie(c, t[j]) = addmul(t[j], x[j], y[i], c);
+        }
+        auto tmp = addc(t[S], c);
+        t[S] = tmp.value;
+        auto d = tmp.carry;
+
+        c = 0;
+        auto m = t[0] * m_mod_inv;
+        std::tie(c, t[0]) = addmul(t[0], m, mod[0], c);
+        for (size_t j = 1; j != S; ++j)
+        {
+            std::tie(c, t[j - 1]) = addmul(t[j], m, mod[j], c);
+        }
+        tmp = addc(t[S], c);
+        t[S - 1] = tmp.value;
+        t[S] = d + tmp.carry;  // TODO: Carry is 0 for sparse modulus.
+    }
+
+    if (t >= mod)  // TODO: cannot overflow if modulus is sparse (e.g. 255 bits).
+        t -= mod;
+
+    return static_cast<UintT>(t);
+}
+
+template <typename UintT>
+UintT ModArith<UintT>::to_mont(const UintT& x) const noexcept
+{
+    return mul(x, m_r_squared);
```

Comment:
Isn't Montgomery form `aR mod N` and not `aR² mod N` ?

Reply:
True, but `mul` is actually `mont_mul` which is `abR⁻¹ mod N`. Therefore: `mul(a, R²)` is `aR²R⁻¹ mod N` is `aR mod N`.

I will add a comment.

Reply:
Montgomery reduction of an input `a` yields `aR**-1 mod N`.  So to convert `a` to Montgomery form, `R²` is passed and the extra factor of `R` cancels out.

Reply:
oh I see! missed that it calls Montgoemry multiplication. Why not call it `mont_mul()` ?

Reply:
The internal representation (e.g. Montgomery form) is abstracted away in EVMMAX. The multiplication is just `MULMODX`. But `to_mont()` and `from_mont()` maps to `LOADX` and `STOREX`. Here we started with add, sub and mul, but we don't have proper load, store yet.